include/deal.II/lac/linear_operator.h

```diff
@@ -2126,6 +2126,289 @@ operator*(const PackagedOperation<Range> &comp,
+            ExcDimensionMismatch(v.n_blocks(), block_matrix.n_block_cols()));
+    Assert( u.n_blocks() == block_matrix.n_block_rows(),
+            ExcDimensionMismatch(u.n_blocks(), block_matrix.n_block_rows()));
+
+    for (unsigned int i = 0; i < block_matrix.n_block_cols(); ++i)
+      {
+        v.block(i) *= 0;
+        for (unsigned int j = 0; j < i; ++j)
+          block_matrix.block(j,i).vmult_add(v.block(i), u.block(j));
+      }
+  };
+
+  return_op.Tvmult_add = [&block_matrix](Domain &v, const Range &u)
+  {
+    Assert( v.n_blocks() == block_matrix.n_block_cols(),
+            ExcDimensionMismatch(v.n_blocks(), block_matrix.n_block_cols()));
+    Assert( u.n_blocks() == block_matrix.n_block_rows(),
+            ExcDimensionMismatch(u.n_blocks(), block_matrix.n_block_rows()));
+
+    for (unsigned int i = 0; i < block_matrix.n_block_cols(); ++i)
+      for (unsigned int j = 0; j < i; ++j)
+        block_matrix.block(j,i).vmult_add(v.block(i), u.block(j));
+  };
+
+  return return_op;
+}
+
+
+/**
+ * @relates LinearOperator
+ *
+ * Let M be a block matrix of the form Id + T made of nxn blocks
+ * and where T is a lower / upper triangular (without diagonal).
+ * Then, its inverse is of the form:
+ * @code
+ *  Id + sum_{i=1}^{n-1} (-1)^i T^i
+ * @endcode
+ * This formula can be used to invert all triangular matrices
+ * (diagonal included of course).
+ *
+ * This function takes a block matrix @p a (possibly full) and a linear block operator
+ * @p inverse_diagonal made of the inverse of the diagonal blocks inverses.
+ * The output is the inverse of the matrix in the case of a triangular matrix and
+ * as inverse_diagonal its diagonal blocks inverses.
+ * Otherwise, the result is a preconditioner.
+ *
+ * The parameter @p lower is a bool that allows to specify if we want
+ * to use lower triangular part of a (true, this is the default value) or
+ * to use upper triangular part of a (false).
+ * @ingroup LAOperators
+ */
+template <typename Range = BlockVector<double>,
+          typename Domain = Range,
+          typename BlockMatrix>
+LinearOperator<Range, Domain>
+preconditioner_from_diagonal_inverse( const BlockMatrix &block_matrix,
+                                      const LinearOperator<Range, Domain> &inverse_diagonal,
+                                      bool lower = true)
+{
+  Assert( block_matrix.n_block_rows() == block_matrix.n_block_cols(),
+          ExcDimensionMismatch(block_matrix.n_block_rows(),block_matrix.n_block_cols()) );
+
+  LinearOperator<Domain, Range> op_a;
+
+  if (lower)
+    {
+      op_a = lower_triangular_operator(block_matrix);
+    }
+  else
+    {
+      op_a = upper_triangular_operator(block_matrix);
+    }
+
+  auto id     = identity_operator(op_a.reinit_range_vector);
+  auto result = identity_operator(op_a.reinit_range_vector);
+
+  for (unsigned int i = 0; i < block_matrix.n_block_cols() - 1; ++i)
+    result = id - inverse_diagonal * op_a * result;
```

Comment:
Maybe i did not understood your observation. 
In the case it concerns math the answer is the following:
We are using matrix with the following assumption: n_block_cols() =  n_block_rows(). This mean these matrices are squared. Moreover, we want to invert a lower / upper matrix without diagonal. Lets name this matrix T. 
T is nilpotent and therefore Id + N^n_block_cols() is equals to Id.
This means (Id - N^n) = (Id + N)(Id - N + ... (-1)^(n-1) N^(n-1))  --> (Id - N)^-1 = Id - N + ... (-1)^(n-1) N^(n-1).
This means that if you have a matrix with 2 blocks x 2 blocks you can invert it in one step.
Generally, you can invert a matrix in one step if it is atomic ( NxN = Id ) otherwise the previous loop  is required. 
Test it with a matrix like 
| 1    0    0 |
| 10  1    0 |
| 10  10  1 | 
This matrix should require 2 two steps.

Sorry for the -vv but I would be as clear as possible.

In the case the question was different, can you explain me the question?